{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cf2aeb1e-0f98-472e-bbc8-9a4ec1878cd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "from gtfparse import read_gtf\n",
    "from Bio import SeqIO\n",
    "import pybedtools\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "03447585-6240-4e57-a2b8-fc972ffbf242",
   "metadata": {},
   "outputs": [],
   "source": [
    "# download hg19.fa file: https://hgdownload.soe.ucsc.edu/goldenPath/hg19/bigZips/hg19.fa.gz\n",
    "# download expression data file: ftp://download.big.ac.cn/lncexpdb/7-CancerCellLines/1-CancerCellLineEncyclopedia/expression_profiles_CCLE.tar.gz\n",
    "# put them in ./data/raw_data/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e97e0468-6b3d-4e30-be0f-1ed4ed4c34a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_gtf_to_bed(anno):\n",
    "    \"\"\" \n",
    "    process gtf (1-base) to bed (0-base), \n",
    "    and key depend on gene type: lncRNA('gene_id'); protein_coding gene ('gene_name')\n",
    "    \"\"\"\n",
    "    anno = anno[['seqname','start','end','gene_name','score','strand','gene_id']].copy()\n",
    "    anno.start = anno.start - 1\n",
    "    anno.score = '.'\n",
    "    \n",
    "    return anno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b0747b18-ba03-4d4f-b256-9a028ec9375b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_gene(file):\n",
    "    \"\"\" \n",
    "    read gene list\n",
    "    \"\"\" \n",
    "    with open(file) as f:\n",
    "        gene = f.readlines()\n",
    "    gene = [x.rstrip() for x in gene]\n",
    "    \n",
    "    return gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "be84826b-535e-4534-8467-af70f74aa9e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "class CancerDriverLnc:\n",
    "    \"\"\"\n",
    "    CADTAD: the pipeline of mining cancer driver lncRNAs\n",
    "    \"\"\"\n",
    "    def __init__(self, gene, gene_type='gene_name'):\n",
    "        self.gene = gene\n",
    "        self.gene_type = gene_type\n",
    "        \n",
    "    def extract_gene_bed(self, gene_gtf):\n",
    "        \"1.extract gene bed\"\n",
    "        gtf = read_gtf(gene_gtf)\n",
    "        if self.gene_type == 'gene_name':\n",
    "            self.gene_gtf = gtf[gtf.gene_name.isin(self.gene)]\n",
    "        elif self.gene_type == 'gene_id':\n",
    "            self.gene_gtf = gtf[gtf.gene_id.isin(self.gene)]\n",
    "        \n",
    "        self.gene_bed = process_gtf_to_bed(self.gene_gtf)\n",
    "        print(\"The number of genes:\", self.gene_bed.shape[0])\n",
    "    \n",
    "    def gene_TAD(self, TAD_file_path):\n",
    "        \"2.gene_TAD\"\n",
    "        TAD_files = os.listdir(TAD_file_path)\n",
    "        TAD_bed = [pybedtools.BedTool(''.join([TAD_file_path,i])) for i in TAD_files]\n",
    "        gene_TAD = {}\n",
    "        df = {'gene_name':[],'gene_id':[],'CDT':[]}\n",
    "        \n",
    "        for i in range(self.gene_bed.shape[0]):\n",
    "            tmp_single_gene_bed = pybedtools.BedTool.from_dataframe(self.gene_bed.iloc[[i],])\n",
    "            tmp_intersect = [j.intersect(tmp_single_gene_bed, wa=True, F=1) for j in TAD_bed]\n",
    "            gene_TAD[list(self.gene_bed.gene_id)[i]] = tmp_intersect[0].cat(*tmp_intersect[1:])\n",
    "            \n",
    "            df['gene_name'].append(list(self.gene_bed.gene_name)[i])\n",
    "            df['gene_id'].append(list(self.gene_bed.gene_id)[i])\n",
    "            interval = tmp_intersect[0].cat(*tmp_intersect[1:]).to_dataframe()\n",
    "            try:\n",
    "                df['CDT'].append(interval.iloc[0,0]+':'+str(interval.iloc[0,1])+'-'+str(interval.iloc[0,2]))\n",
    "            except:\n",
    "                df['CDT'].append(None)\n",
    "        \n",
    "        self.gene_TAD = gene_TAD\n",
    "        self.df = pd.DataFrame(df)\n",
    "        \n",
    "        print(\"The number of genes having TAD:\", sum(-self.df.CDT.isnull()))\n",
    "        print(\"The number of Cancer Driver TADs:\", len(set(self.df.CDT)) - list(set(self.df.CDT)).count(None))\n",
    "        \n",
    "    \n",
    "    def gene_TAD_lncRNA(self,lnc_gtf):\n",
    "        \"3.gene_TAD_lncRNA\"\n",
    "        lnc_gtf = read_gtf(lnc_gtf)\n",
    "        print(\"total lncRNA:\",lnc_gtf.shape[0])\n",
    "        lnc_bed = process_gtf_to_bed(lnc_gtf)\n",
    "        lnc_bed_py = pybedtools.BedTool.from_dataframe(lnc_bed)\n",
    "        \n",
    "        gene_TAD_lnc = []\n",
    "        \n",
    "        for i in self.gene_TAD.keys():\n",
    "            tmp_TAD = self.gene_TAD[i]\n",
    "            try:\n",
    "                TAD_lnc = list(tmp_TAD.intersect(lnc_bed_py, F=1, wo=True).to_dataframe().iloc[:,9])\n",
    "            except:\n",
    "                TAD_lnc = None\n",
    "            \n",
    "            gene_TAD_lnc.append(TAD_lnc)\n",
    "        \n",
    "        self.df['lnc'] = gene_TAD_lnc\n",
    "        self.df = self.df.dropna(axis=0)\n",
    "        \n",
    "        \n",
    "        lnc_n = []\n",
    "        for i in self.df.lnc:\n",
    "            lnc_n = lnc_n + i\n",
    "        \n",
    "        self.lnc_bed = lnc_bed[lnc_bed.gene_id.isin(list(set(lnc_n)))]\n",
    "        print(\"Cancer Driver TADs contain\", len(set(lnc_n)), \"lncRNAs\")\n",
    "        print(\"Total\", sum([len(i) for i in self.df.lnc]), \"gene-lncRNA pairs\")\n",
    "        \n",
    "        # 新加的一行，为了跳过bind去算cor\n",
    "        self.df_TAD = pd.DataFrame(columns=['gene_id','lnc_id','gene_name','CDT'])\n",
    "        for i in self.df.index:\n",
    "            gene_id = self.df.loc[i,'gene_id']\n",
    "            gene_name = self.df.loc[i,'gene_name']\n",
    "            CDT= self.df.loc[i,'CDT']\n",
    "            lnc = self.df.loc[i,'lnc']\n",
    "            \n",
    "            length = len(lnc)\n",
    "            \n",
    "            temp_df_TAD = pd.DataFrame({'gene_id':[gene_id]*length, 'lnc_id':lnc, 'gene_name':[gene_name]*length, 'CDT':[CDT]*length})\n",
    "            self.df_TAD = pd.concat([self.df_TAD, temp_df_TAD], ignore_index=True)\n",
    "            \n",
    "            \n",
    "            \n",
    "    \n",
    "    # 4.bind\n",
    "    def get_fasta(self, fasta_file, outpath):\n",
    "        fasta = pybedtools.BedTool(fasta_file)\n",
    "        gene_bed = self.gene_bed[self.gene_bed.gene_id.isin(list(self.df.gene_id))]\n",
    "        gene_bed.gene_name = gene_bed.gene_id\n",
    "        lnc_bed = self.lnc_bed\n",
    "        lnc_bed.gene_name = lnc_bed.gene_id\n",
    "        \n",
    "        if os.path.exists(outpath + 'gene_fa') or os.path.exists(outpath + 'lnc_fa'):\n",
    "            print('gene_fa or lnc_fa exists!')\n",
    "        else:\n",
    "            os.mkdir(''.join([outpath,'gene_fa']))\n",
    "            os.mkdir(''.join([outpath,'lnc_fa']))\n",
    "        \n",
    "            for i in range(self.df.shape[0]):\n",
    "                tmp_gene_bed = pybedtools.BedTool.from_dataframe(gene_bed[gene_bed.gene_id == list(self.df.gene_id)[i]])\n",
    "                tmp_gene_fa = tmp_gene_bed.sequence(fi=fasta, nameOnly=True, s=True)\n",
    "                tmp_lnc_bed = pybedtools.BedTool.from_dataframe(lnc_bed[lnc_bed.gene_id.isin(list(self.df.lnc)[i])])\n",
    "                tmp_lnc_fa = tmp_lnc_bed.sequence(fi=fasta, nameOnly=True, s=True)\n",
    "                \n",
    "                with open(''.join([outpath,'gene_fa/',list(self.df.gene_id)[i],'.fa']),'w') as f:\n",
    "                    f.write(open(tmp_gene_fa.seqfn).read().upper())\n",
    "            \n",
    "                with open(''.join([outpath,'lnc_fa/', list(self.df.gene_id)[i],'.fa']),'w') as f:\n",
    "                    f.write(open(tmp_lnc_fa.seqfn).read().upper())\n",
    "        \n",
    "        \n",
    "    def triplexator(self, script_path, gene_path, data_path):\n",
    "        os.system(\"cat %s | /home/rzy/software/rush -j 5 'bash %s {1} %s'\" % (gene_path, script_path, data_path))\n",
    "    \n",
    "    def QmRLFSfinder(self, script_path, gene_path, data_path, out_dir):\n",
    "        os.system(\"cat %s | /home/rzy/software/rush -j 20 'bash %s {1} %s %s'\" % (gene_path, script_path, data_path, out_dir))\n",
    "    \n",
    "    def QmRLFSfinder_results_process(self, gene_name, out_dir):\n",
    "        try:\n",
    "            lnc_path = [out_dir, 'QmRLFS-finder/', gene_name, \".out.table.txt\"]\n",
    "            lnc = pd.read_table((\"\").join(lnc_path))\n",
    "        except:\n",
    "            print(gene_name,\"no R-loop data\")\n",
    "        else:\n",
    "            lnc_seq = lnc[lnc.strand == \"+\"].sequence_RIZ\n",
    "            lnc_name = lnc[lnc.strand == \"+\"]['#Name']\n",
    "        \n",
    "            gene_path = [out_dir,\"gene_fa/\",gene_name,\".fa\"]\n",
    "            gene = SeqIO.read((\"\").join(gene_path),\"fasta\")\n",
    "        \n",
    "            num1 = [] # +\n",
    "            num2 = [] # -\n",
    "            for lnc in set(lnc_name):\n",
    "                bind_sites1 = 0\n",
    "                bind_sites2 = 0\n",
    "                for seq in lnc_seq[lnc_name == lnc]:\n",
    "                    bind_sites1 = bind_sites1 + gene.seq.count(seq)\n",
    "                    bind_sites2 = bind_sites2 + gene.seq.reverse_complement().count(seq)\n",
    "                num1.append(bind_sites1)\n",
    "                num2.append(bind_sites2)\n",
    "\n",
    "    \n",
    "            # 将以上得到的结果整理成一个gene、+_bind_sites、-_bind_sites、all_bind_sites开头的表格\n",
    "            data = {\"gene_id\":[gene_name]*len(set(lnc_name)),\n",
    "                    \"lnc_id\":list(set(lnc_name)),\n",
    "               \"+_bind_sites\":num1,\n",
    "               \"-_bind_sites\":num2,\n",
    "               \"all_bind_sites\":pd.Series(num1) + pd.Series(num2)}\n",
    "            results = pd.DataFrame(data, columns = [\"gene_id\",\"lnc_id\",\"+_bind_sites\",\"-_bind_sites\",\"all_bind_sites\"])\n",
    "    \n",
    "            return(results)\n",
    "    \n",
    "    def gene_lncRNA_bind(self, fasta_file, outpath='./data/pipeline/bind/'):\n",
    "        # fasta process\n",
    "        self.get_fasta(fasta_file, outpath)\n",
    "        \n",
    "        # get gene_id\n",
    "        self.df.gene_id.to_csv(outpath+'gene_id.txt', index=False, header=None)\n",
    "        \n",
    "        # triplexator\n",
    "        raw_path = os.getcwd()\n",
    "        script_path_1 = ''.join([raw_path,'/triplexator.sh'])\n",
    "        \n",
    "        os.chdir(outpath)\n",
    "        data_path = os.getcwd()\n",
    "        \n",
    "        if os.path.exists(data_path+'/Triplexator'):\n",
    "            print('Triplexator exists!')\n",
    "        else:\n",
    "            os.mkdir('./Triplexator')\n",
    "            os.chdir('./Triplexator')\n",
    "            #triplex_res_path = os.getcwd()\n",
    "            \n",
    "            self.triplexator(script_path_1, data_path+'/gene_id.txt',data_path)\n",
    "        os.chdir(raw_path)\n",
    "\n",
    "        # QmRLFS-finder\n",
    "        script_path_2 = ''.join([raw_path,'/QmRLFS-finder.sh'])\n",
    "        \n",
    "        if os.path.exists(data_path+'/QmRLFS-finder'):\n",
    "            print('QmRLFS-finder exists!')\n",
    "        else:\n",
    "            os.chdir(outpath)\n",
    "            os.mkdir('./QmRLFS-finder')\n",
    "            os.chdir('./QmRLFS-finder')\n",
    "        \n",
    "            out_dir = os.getcwd()\n",
    "        \n",
    "            self.QmRLFSfinder(script_path_2, data_path+'/gene_id.txt', data_path, out_dir)\n",
    "        os.chdir(raw_path)\n",
    "        \n",
    "        \n",
    "    def gene_lncRNA_bind_porcess(self, outpath='./data/pipeline/bind/'):\n",
    "        # triplexator results process\n",
    "        if os.path.exists(outpath+'triplextor_results.txt'):\n",
    "            print('triplextor_results.txt exists!')\n",
    "        else:\n",
    "            with open(outpath+'triplextor_results.txt','w') as f:\n",
    "                file_list = os.listdir(outpath+'Triplexator')\n",
    "                matchPattern = re.compile(r'#')\n",
    "                for i in file_list:\n",
    "                    file = open(outpath+'Triplexator/'+i+'/'+i+'.tpx.summary')\n",
    "                    lineList = []\n",
    "                    while 1:\n",
    "                        line = file.readline()\n",
    "                        if not line:\n",
    "                            break\n",
    "                        elif matchPattern.search(line):\n",
    "                            pass\n",
    "                        else:\n",
    "                            lineList.append(line)\n",
    "                    file.close()\n",
    "                \n",
    "                    for j in lineList:\n",
    "                        f.writelines(j)\n",
    "        \n",
    "        os.system('wc -l %s' % (outpath+'triplextor_results.txt'))\n",
    "        self.triplextor_results = pd.read_table(outpath+'triplextor_results.txt', \n",
    "                                               names=['gene_id','lnc_id','Total (abs)', 'Total (rel)', 'GA (abs)', 'GA (rel)', 'TC (abs)', 'TC (rel)', 'GT (abs)', 'GT (rel)', 'other'])\n",
    "        self.triplextor_results['gene_id'] = [i.split('(')[0] for i in self.triplextor_results['gene_id']]\n",
    "        self.triplextor_results['lnc_id'] = [i.split('(')[0] for i in self.triplextor_results['lnc_id']]\n",
    "        self.triplextor_results = pd.merge(self.triplextor_results, self.df.iloc[:,0:3], on='gene_id')\n",
    "        \n",
    "        # QmRLFS-finder results process\n",
    "        if os.path.exists(outpath+'QmRLFS-finder_results.txt'):\n",
    "            print('QmRLFS-finder_results.txt exists!')\n",
    "        else:\n",
    "            inCDT_lnc =  pd.DataFrame(columns = [\"gene_id\",\"lnc_id\",\"+_bind_sites\",\"-_bind_sites\",\"all_bind_sites\"])\n",
    "            for i in self.df.gene_id:\n",
    "                inCDT_lnc = inCDT_lnc.append(self.QmRLFSfinder_results_process(i, outpath), ignore_index=True)\n",
    "        \n",
    "            inCDT_lnc[inCDT_lnc.all_bind_sites != 0].to_csv(outpath + \"QmRLFS-finder_results.txt\",sep=\"\\t\",index=False)\n",
    "        \n",
    "        os.system('wc -l %s' % (outpath+'QmRLFS-finder_results.txt'))\n",
    "        self.QmRLFSfinder_results = pd.read_table(outpath+'QmRLFS-finder_results.txt')\n",
    "        self.QmRLFSfinder_results['lnc_id'] = [i.split('(')[0] for i in self.QmRLFSfinder_results['lnc_id']]\n",
    "        self.QmRLFSfinder_results = pd.merge(self.QmRLFSfinder_results, self.df.iloc[:,0:3], on='gene_id')\n",
    "        \n",
    "        self.df_TAD_bind = pd.merge(self.triplextor_results[['gene_id','lnc_id','Total (abs)', 'Total (rel)', 'gene_name', 'CDT']], \n",
    "                                    self.QmRLFSfinder_results[['gene_id','lnc_id','all_bind_sites', 'gene_name', 'CDT']],\n",
    "                                    on=['gene_id','lnc_id','gene_name', 'CDT'], how='outer').fillna(0)\n",
    "    \n",
    "    # 5.cor\n",
    "    def coexpression(self, exp_samples, trans=True, bind=True):\n",
    "        \"exp_samples: expression matrix, columns are genes, rows are samples. trans: if columns are samples, rows are genes.\"\n",
    "        exp = pd.read_table(exp_samples, index_col=0)\n",
    "        if trans:\n",
    "            exp = exp.T\n",
    "            \n",
    "        if bind:\n",
    "            df_TAD_bind_cor = self.df_TAD_bind.copy()\n",
    "            df_TAD_bind_cor['pearson_cor'] = 0\n",
    "            df_TAD_bind_cor['spearman_cor'] = 0\n",
    "        \n",
    "            for i in range(df_TAD_bind_cor.shape[0]):\n",
    "                df_TAD_bind_cor.iloc[i,7] = exp[df_TAD_bind_cor.iloc[i,0]].corr(exp[df_TAD_bind_cor.iloc[i,1]], method='pearson')\n",
    "                df_TAD_bind_cor.iloc[i,8] = exp[df_TAD_bind_cor.iloc[i,0]].corr(exp[df_TAD_bind_cor.iloc[i,1]], method='spearman')\n",
    "                \n",
    "            self.df_TAD_bind_cor = df_TAD_bind_cor.copy()\n",
    "        else:\n",
    "            df_TAD_cor = self.df_TAD.copy()\n",
    "            df_TAD_cor['pearson_cor'] = 0\n",
    "            df_TAD_cor['spearman_cor'] = 0\n",
    "        \n",
    "            for i in range(df_TAD_cor.shape[0]):\n",
    "                df_TAD_cor.iloc[i,4] = exp[df_TAD_cor.iloc[i,0]].corr(exp[df_TAD_cor.iloc[i,1]], method='pearson')\n",
    "                df_TAD_cor.iloc[i,5] = exp[df_TAD_cor.iloc[i,0]].corr(exp[df_TAD_cor.iloc[i,1]], method='spearman')\n",
    "                \n",
    "            \n",
    "            self.df_TAD_cor = df_TAD_cor.copy()\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c16e5fba-efdc-4870-8975-8a7e73381c66",
   "metadata": {},
   "outputs": [],
   "source": [
    "cosmic = read_gene('./data/raw_data/cosmic.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9e619bb9-50c5-4009-bddc-ef423f84988f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_type', 'gene_name', 'level', 'hgnc_id', 'havana_gene', 'tag']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of genes: 556\n",
      "The number of genes having TAD: 536\n",
      "The number of Cancer Driver TADs: 474\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_type', 'gene_classification', 'type', 'gene_name']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total lncRNA: 94810\n",
      "Cancer Driver TADs contain 30289 lncRNAs\n",
      "Total 46701 gene-lncRNA pairs\n"
     ]
    }
   ],
   "source": [
    "pancancer = CancerDriverLnc(cosmic,'gene_name')\n",
    "pancancer.extract_gene_bed('./data/raw_data/protein_coding_gene.gtf')\n",
    "pancancer.gene_TAD('./data/raw_data/TAD/solid_T/')\n",
    "pancancer.gene_TAD_lncRNA('./data/raw_data/lncRNA_gene.gtf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b962b915-bc8f-4367-a605-c05920ceaed1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene_fa or lnc_fa exists!\n",
      "Triplexator exists!\n",
      "QmRLFS-finder exists!\n"
     ]
    }
   ],
   "source": [
    "pancancer.gene_lncRNA_bind('./data/raw_data/hg19.fa', outpath='./data/pipeline/bind/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5bb6b314-eba3-4b05-838f-596073318d00",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "triplextor_results.txt exists!\n",
      "6740 ./data/pipeline/bind/triplextor_results.txt\n",
      "QmRLFS-finder_results.txt exists!\n",
      "5224 ./data/pipeline/bind/QmRLFS-finder_results.txt\n"
     ]
    }
   ],
   "source": [
    "pancancer.gene_lncRNA_bind_porcess()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "977e4107-41fb-4bb6-8e67-af1955d1167c",
   "metadata": {},
   "outputs": [],
   "source": [
    "pancancer.coexpression('./data/raw_data/CCLE_geneTPM.tsv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "501b74a5-cf74-4489-9397-715097503be7",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>lnc_id</th>\n",
       "      <th>Total (abs)</th>\n",
       "      <th>Total (rel)</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>CDT</th>\n",
       "      <th>all_bind_sites</th>\n",
       "      <th>pearson_cor</th>\n",
       "      <th>spearman_cor</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000171735.19</td>\n",
       "      <td>HSALNG0000544</td>\n",
       "      <td>30.0</td>\n",
       "      <td>7.230000e-10</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>chr1:6760000-8040000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.001363</td>\n",
       "      <td>0.105666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000171735.19</td>\n",
       "      <td>HSALNG0142097</td>\n",
       "      <td>14.0</td>\n",
       "      <td>1.790000e-10</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>chr1:6760000-8040000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.015401</td>\n",
       "      <td>-0.011828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000171735.19</td>\n",
       "      <td>HSALNG0000546</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.640000e-11</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>chr1:6760000-8040000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.006084</td>\n",
       "      <td>0.037206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000171735.19</td>\n",
       "      <td>HSALNG0000547</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.210000e-11</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>chr1:6760000-8040000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.441343</td>\n",
       "      <td>0.382724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000171735.19</td>\n",
       "      <td>HSALNG0000548</td>\n",
       "      <td>184.0</td>\n",
       "      <td>1.750000e-09</td>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>chr1:6760000-8040000</td>\n",
       "      <td>82.0</td>\n",
       "      <td>0.006662</td>\n",
       "      <td>0.037073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10204</th>\n",
       "      <td>ENSG00000147403.16</td>\n",
       "      <td>HSALNG0140697</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>RPL10</td>\n",
       "      <td>chrX:150880000-154200000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.035880</td>\n",
       "      <td>-0.117370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10205</th>\n",
       "      <td>ENSG00000147403.16</td>\n",
       "      <td>HSALNG0151156</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>RPL10</td>\n",
       "      <td>chrX:150880000-154200000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-0.037283</td>\n",
       "      <td>-0.108416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10206</th>\n",
       "      <td>ENSG00000147403.16</td>\n",
       "      <td>HSALNG0140838</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>RPL10</td>\n",
       "      <td>chrX:150880000-154200000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10207</th>\n",
       "      <td>ENSG00000147403.16</td>\n",
       "      <td>HSALNG0140809</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>RPL10</td>\n",
       "      <td>chrX:150880000-154200000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.330483</td>\n",
       "      <td>0.246601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10208</th>\n",
       "      <td>ENSG00000214827.11</td>\n",
       "      <td>HSALNG0140897</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>MTCP1</td>\n",
       "      <td>chrX:154120000-154520000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.078262</td>\n",
       "      <td>0.135518</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10209 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  gene_id         lnc_id  Total (abs)   Total (rel) gene_name  \\\n",
       "0      ENSG00000171735.19  HSALNG0000544         30.0  7.230000e-10    CAMTA1   \n",
       "1      ENSG00000171735.19  HSALNG0142097         14.0  1.790000e-10    CAMTA1   \n",
       "2      ENSG00000171735.19  HSALNG0000546          1.0  1.640000e-11    CAMTA1   \n",
       "3      ENSG00000171735.19  HSALNG0000547          4.0  3.210000e-11    CAMTA1   \n",
       "4      ENSG00000171735.19  HSALNG0000548        184.0  1.750000e-09    CAMTA1   \n",
       "...                   ...            ...          ...           ...       ...   \n",
       "10204  ENSG00000147403.16  HSALNG0140697          0.0  0.000000e+00     RPL10   \n",
       "10205  ENSG00000147403.16  HSALNG0151156          0.0  0.000000e+00     RPL10   \n",
       "10206  ENSG00000147403.16  HSALNG0140838          0.0  0.000000e+00     RPL10   \n",
       "10207  ENSG00000147403.16  HSALNG0140809          0.0  0.000000e+00     RPL10   \n",
       "10208  ENSG00000214827.11  HSALNG0140897          0.0  0.000000e+00     MTCP1   \n",
       "\n",
       "                            CDT  all_bind_sites  pearson_cor  spearman_cor  \n",
       "0          chr1:6760000-8040000             0.0    -0.001363      0.105666  \n",
       "1          chr1:6760000-8040000             0.0    -0.015401     -0.011828  \n",
       "2          chr1:6760000-8040000             0.0    -0.006084      0.037206  \n",
       "3          chr1:6760000-8040000             0.0     0.441343      0.382724  \n",
       "4          chr1:6760000-8040000            82.0     0.006662      0.037073  \n",
       "...                         ...             ...          ...           ...  \n",
       "10204  chrX:150880000-154200000             3.0     0.035880     -0.117370  \n",
       "10205  chrX:150880000-154200000             1.0    -0.037283     -0.108416  \n",
       "10206  chrX:150880000-154200000             2.0          NaN           NaN  \n",
       "10207  chrX:150880000-154200000             2.0     0.330483      0.246601  \n",
       "10208  chrX:154120000-154520000             1.0     0.078262      0.135518  \n",
       "\n",
       "[10209 rows x 9 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pancancer.df_TAD_bind_cor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "66f74204-9b2d-47f6-a6db-6b34e9183925",
   "metadata": {},
   "outputs": [],
   "source": [
    "cosmic_gene_info = pd.read_table('../0.raw_data/cosmic_genes_info.txt',sep=' ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0ce3e160-2eda-4d2e-b158-50a9255a889b",
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer_driver_lncRNA = pd.merge(pancancer.df_TAD_bind_cor[np.abs(pancancer.df_TAD_bind_cor['spearman_cor']) > 0.5], cosmic_gene_info, on='gene_name')\n",
    "\n",
    "CDL_type = cancer_driver_lncRNA.groupby(['lnc_id'])[['TSG','OG']].sum()\n",
    "CDL_type['type'] = 'CDL'\n",
    "CDL_type['type'][CDL_type['OG'] == 0] = 'Tumor suppressive lncRNA'\n",
    "CDL_type['type'][CDL_type['TSG'] == 0] = 'Oncogenic lncRNA'\n",
    "CDL_type['type'][(CDL_type['OG'] != 0) & (CDL_type['TSG'] != 0)]= 'Dual lncRNA'\n",
    "CDL_type = pd.DataFrame({'lnc_id':list(CDL_type.index),'lnc_type':list(CDL_type.type)})\n",
    "\n",
    "cancer_driver_lncRNA = pd.merge(cancer_driver_lncRNA, CDL_type, on='lnc_id')\n",
    "#cancer_driver_lncRNA.to_csv('./data/pipeline/cancer_driver_lncRNAs.txt', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d13b10fd-d662-4208-8b35-a57a385f5ea9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Oncogenic lncRNA            256\n",
       "Tumor suppressive lncRNA    177\n",
       "Dual lncRNA                  75\n",
       "Name: lnc_type, dtype: int64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CDL_type.drop_duplicates()['lnc_type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd4d8d52-a8d5-4a83-93cd-6a81fe54c20e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
